src/map_inner.rs

```diff
@@ -0,0 +1,1142 @@
+#![allow(clippy::indexing_slicing)] // indexing will never panic here, because all of the index has been mod by the length of the vector.
+
+use std::collections::hash_map::RandomState;
+
+use utilities::{Cast, OverflowArithmetic};
+
+use std::hash::{BuildHasher, Hash, Hasher};
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use super::pointer::{AtomicPtr, SharedPtr};
+
+// Re-export `crossbeam_epoch::pin()` and `crossbeam_epoch::Guard`.
+pub use crossbeam_epoch::{pin, Guard};
+
+/// `KVPair` contains the key-value pair.
+#[derive(Debug)]
+pub struct KVPair<K, V> {
+    // TODO: maybe cache both hash keys here.
+    /// `key` is the key of a KV pair.
+    pub key: K,
+    /// `value` is the value of a KV pair.
+    pub value: V,
+}
+
+/// `SlotIndex` represents the index of a slot inside the hashtable.
+/// The slot index is composed by `tbl_idx` and `slot_idx`.
+#[derive(Clone, Copy, Debug)]
+struct SlotIndex {
+    /// `tbl_idx` is the index of the table.
+    tbl_idx: usize,
+    /// `slot_idx` is the index of the slot inside one table.
+    slot_idx: usize,
+}
+
+/// `SlotState` represents the state of a slot.
+/// A slot could be in one of the four states: null, key, reloc and copied.
+#[derive(PartialEq)]
+enum SlotState {
+    /// `NullOrKey` means a slot is empty(null) or is occupied by a key-value
+    /// pair normally without any other flags.
+    NullOrKey,
+    /// `Reloc` means a slot is being relocated to the other slot.
+    Reloc,
+    /// `Copied` means a slot is being copied to the new map during resize or
+    /// has been copied to the new map.
+    Copied,
+}
+
+impl SlotState {
```

Comment:
`impl From for SlotState`

when you implement From trait, you get Into trait implemented for free.